fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -118,81 +153,137 @@ public static byte[] decode(@Nullable byte[] data) {
+                    LuceneLogMessageKeys.COMPRESSED_EVENTUALLY, state.isCompressed(),
+                    LuceneLogMessageKeys.ENCRYPTED_EVENTUALLY, state.isCompressed(),
                     LuceneLogMessageKeys.ENCODED_DATA_SIZE, data.length,
                     LuceneLogMessageKeys.ORIGINAL_DATA_SIZE, decoded.length));
         }
+
         return decoded;
     }
 
     @Nonnull
-    private static byte[] compressIfNeeded(boolean compressionNeeded, @Nonnull byte[] data, @Nonnull ByteBuffersDataOutput encodedDataOutput,
-                                           @Nonnull EncodingState state, int offset) {
-        if (!compressionNeeded) {
-            return fallBackToUncompressed(encodedDataOutput.toArrayCopy(), data, state, offset);
+    private static byte[] compressIfNeeded(@Nonnull EncodingState state, @Nonnull ByteBuffersDataOutput encodedDataOutput,
+                                           @Nonnull byte[] uncompressedData, int prefixLength)
+            throws IOException {
+        if (!state.isCompressed()) {
+            return fallBackToUncompressed(state, uncompressedData, encodedDataOutput.toArrayCopy(), prefixLength);
         }
 
         try (Compressor compressor = CompressionMode.HIGH_COMPRESSION.newCompressor()) {
             encodedDataOutput.writeByte(COMPRESSION_VERSION_FOR_HIGH_COMPRESSION);
-            encodedDataOutput.writeVInt(data.length);
-            compressor.compress(data, 0, data.length, encodedDataOutput);
+            encodedDataOutput.writeVInt(uncompressedData.length);
+            compressor.compress(uncompressedData, 0, uncompressedData.length, encodedDataOutput);
             final byte[] compressedData = encodedDataOutput.toArrayCopy();
 
             // Compress only if it helps to shorten the bytes
-            if (compressedData.length < data.length) {
+            if (compressedData.length < uncompressedData.length) {
                 state.setCompressed(true);
                 return compressedData;
             } else {
-                return fallBackToUncompressed(compressedData, data, state, offset);
+                return fallBackToUncompressed(state, uncompressedData, compressedData, prefixLength);
             }
-        } catch (Exception e) {
-            throw new RecordCoreException("Lucene data compression failure", e);
         }
     }
 
-    private static byte[] fallBackToUncompressed(@Nonnull byte[] compressedData, @Nonnull byte[] originalData,
-                                                 @Nonnull EncodingState state, int offset) {
+    private static byte[] fallBackToUncompressed(@Nonnull EncodingState state, @Nonnull byte[] originalData,
+                                                 @Nonnull byte[] encodedData, int prefixLength) {
+        final byte[] encoded = new byte[originalData.length + prefixLength];
+        System.arraycopy(encodedData, 0, encoded, 0, prefixLength);
+        encoded[0] &= ~ENCODING_COMPRESSED;
```

Comment:
Is this guaranteed to be in byte`[0]` (i.e. what if `prefixLength>1`)

Reply:
Yes, I believe so. Varints are 7-bit little-endian, with a "continuation" bit to know when to stop. So, the compressed flag bit, which is in the low bits, is always in the first byte, no matter how many higher bits there are in the following bytes.